### DISCUSSION TITLE
RLS Policy not working to Update table

### DISCUSSION THREAD

**Discussion by brentrobbins** [supabase/supabase#5097] [General]
Hello, I am trying to create a Policy for a `Profiles` table, but everything I tried doesn't work. `SELECT` and `INSERT` work just fine, but even when I use just `true` for `UPDATE`, but that doesn't work. The user that is connected to that Profiles table works fine even without a specific Update policy. 

What I am trying to do is allow the Profile "owner" and also users with the admin role to edit any/all profiles. Is this possible?

The `profiles.id` foreign key relation is connected with the `auth.users.id` table. 

I really appreciate the help!

**Comment by GaryAustin1** [comment]
You don't provide much info on what your RLS policies are, but a couple of things to check:

Make sure your Update has a using policy,
Make sure your Select policy allows the admins access OR modify your client update call to use parameter { returning: "minimal"} to not also do a read (select).

**Reply by brentrobbins** [reply]
Hey @GaryAustin1, sorry for for adding those details and thanks for the reminder about the parameter. So I added that to my upsert
```
let { error } = await supabase.from('profiles').upsert(updates, { returning: 'minimal' })
```
and I am still receiving the following error `{"message":"new row violates row-level security policy for table \"profiles\"","code":"42501","hint":null,"details":null}`

Here's a screenshot of my current (loose for testing) Policies or the Profiles table:
<img width="935" alt="Screen Shot 2022-01-21 at 6 09 03 PM" src="https://user-images.githubusercontent.com/727491/150618571-f4a4ee01-1514-4d64-b00f-1097f529513c.png">

**Reply by GaryAustin1** [reply]
You have select policy true so no need for minimal.
You had said UPDATE in original post, but here are using UPSERT which is INSERT with option to replace for duplicates...  Your insert policy is only for id match...  UPSERT is translated to  INSERT on CONFLICT do UPDATE so needs to worry about both policies depending on if row exists or not.

Edit:  I'm not sure how you plan to detect an admin, but you will need to add the check for them as an OR in your insert and/or update policies depending on what you are needing.  Not sure you really want upsert for admins if the user already exists.  Probably just update.

**Reply by brentrobbins** [reply]
Yea.... 🤦‍♂️ So I updated it to the following for Select and Update and it seems to work for an admin and the owner.
```
((EXISTS ( SELECT profiles_1.id, profiles_1.role 
FROM profiles profiles_1 
WHERE (profiles_1.role = 'admin'::text))) 
OR (uid() = id))
```

I am sure there's a better way to write it.

**Reply by brentrobbins** [reply]
Well I tested this after temporarily removing the role protection on the edit route and a non admin can actually edit someone else's profile, so somethings not right.

**Reply by brentrobbins** [reply]
OK, I updated it to the following for both Select and Insert and it seems to be working properly, but definitely open other approaches. 

```
( uid() IN (
      SELECT id FROM profiles
  WHERE (profiles.role = 'admin'::text)
)) OR (uid() = id)
```

**Reply by brentrobbins** [reply]
@GaryAustin1 here's a screenshot of the profile's table and the `profiles.id` has a foreign key relation to `auth.users.id`

<img width="1032" alt="Screen Shot 2022-01-21 at 7 29 59 PM" src="https://user-images.githubusercontent.com/727491/150621518-93ffa5e4-d053-49b9-bf51-37c8e9d57417.png">

**Reply by GaryAustin1** [reply]
If this is a large table and used often, it might be best to have a function (with STABLE in the definition) to get the user's admin role and return true or false and use the function in the RLS.   RLS is run on each row when hit.  BUT I don't think you are at the point to worry about optimizing... just get it to work.

Edit:  Ignore function optimization above as that is more for selects... I'd probably do uid()=id first as that should be pretty fast and if it is true, postgres should not run the second part with the query... although it is supposedly pretty smart about optimizing what order to do things, so who knows.

**Reply by brentrobbins** [reply]
Sweet, thanks for your help @GaryAustin1 I really appreciate the help. I will switch `uid()=id` to the beginning.

**Comment by MichalNemec** [comment]
Hello just stumpled upon this too:
```
(auth.uid() IN ( SELECT profiles.id
   FROM profiles
  WHERE (profiles.role = 'admin'::text)))
```

its completely ignored (empty profiles table) and still goes through (posting curl with anon key

exactly same thing for insert.

i would like to do on products table only read.. no insert/update/delete, but its ignored.

**Reply by GaryAustin1** [reply]
@MichalNemec 
It is hard to tell if what you have is related.  If you put the policies as false for insert/update does that work?  Making sure you have your general using/with check logic correct.

**Comment by ThatGuySam** [comment]
I was stuck on this for 2 days. It turns out it's because I was running Supabase in a Node test environment, and Supabase silently fails to setup a session and user id when not in a browser environment like Chromium or jsdom. 

You can fix it by using a browser environment like jsdom for your testing environment or just using Playwright. 

Rough example: 
```js
// @vitest-environment jsdom

test('use jsdom in this test file', () => {
  const expectedName = 'Sam';

  await supabase.from("profiles").insert([{
    id: uuid,
    first_name: expectedName
  }]);

  const { data: profile } = await supabase.from("profiles")
    .select()
    .eq( 'id', uuid )


  expect( profile.first_name ).toEqual( expectedName )
});
```

In Vitest: 
https://vitest.dev/config/#environment

In Jest:
https://jestjs.io/docs/configuration#testenvironment-string

**Reply by ibanks42** [reply]
I've been at this for so many hours at this point. I hadn't considered this in any way. Thank you so much.

**Reply by ThatGuySam** [reply]
@isaiah-banks Glad it helped!
